include/revng/Pipeline/Rank.h

```diff
@@ -17,30 +18,96 @@ class Rank : public DynamicHierarchy<Rank> {
   Rank(llvm::StringRef Name, Rank &Parent) : DynamicHierarchy(Name, Parent) {}
 };
 
-  template<StringTemplateParameter Str, typename LocationType, typename ParentRnk>
-  class RankDefinition : public Rank {
+namespace detail {
+
+template<typename... TypesToAppend, typename... InitialTupleTypes>
+inline constexpr auto appendImpl(std::tuple<InitialTupleTypes...> &&Tuple) {
+  return std::tuple<InitialTupleTypes..., TypesToAppend...>();
+}
+
+template<typename Tuple, typename... TypesToAppend>
+struct AppendToTupleHelper {
+  using type = decltype(appendImpl<TypesToAppend...>(std::declval<Tuple>()));
+};
+
+template<typename First, typename Second>
+struct CH {
+  template<std::size_t Index>
+  constexpr bool operator()() {
+    return std::is_convertible_v<std::tuple_element_t<Index, First>,
+                                 std::tuple_element_t<Index, Second>>;
+  }
+};
+template<typename Tuple>
+inline constexpr auto TS = std::tuple_size_v<Tuple>;
+
+// clang-format off
+template<typename LHS, typename RHS>
+concept TupleConvertibilityCheck = TS<LHS> == TS<RHS> &&
+                                   constexprAnd<TS<LHS>>(CH<LHS, RHS>{});
+// clang-format on
+
+} // namespace detail
+
+template<StringTemplateParameter Str, typename LocationType, typename ParentRnk>
+class RankDefinition : public Rank {
 public:
   static constexpr std::string_view RankName = Str;
   using Type = LocationType;
   using Parent = ParentRnk;
 
   using Definition = RankDefinition<Str, Type, Parent>;
+  using RankTag = Definition;
```

Comment:
I'm a bit confused by this intermediate layer? Can we prune some of them?

Reply:
Yeah, I hate relying on `Tag`s as well. It feels so old-c++-like. But also, this is a simplest way to actually ensure that the object we're getting is indeed an object we're expecting.

What's your call? Should I make external concepts for those instead (the downside I don't like is that if someone is going to modify one of those, they would need to modify both the template and its concept). Or maybe you know a better way to "fingerprint" a template?

Reply:
Ah, now I understand. If you just want a marker maybe use a `constexpr bool IsWhatever = true`, assigning it to `Definition` seems to carry some other semantic meaning, which is a bit confusing.

In general, it depends what guarantees you want to offer? Maybe inheritance from `Rank`? Or the existence of certain members?

Reply:
Members, actually. In essence, it is supposed to guarantee that all the following `typename Parent::.*` are valid. It's easily reproducible with a dedicated concept BUT it does feel like all the relevant names are mentioned twice then.

Reply:
Eh, but if you see `RankDefinition` as an implementation detail and the `concept` as the real thing, it makes sense.

Up to you, the tag is OK too, just make a `bool` maybe.

Reply:
The tag is bool now. Also there's a concept looking for said tag.